### ISSUE TITLE
OnceCycleLR Scheduler computes wrong learning rate in last step

### ISSUE THREAD

**Issue by braun-steven** [pytorch/pytorch#91765]
### 🐛 Describe the bug

The `OneCycleLR` Scheduler seems to perform a wrong last step (sudden increase back to the LR at step `total_steps-2`) when computing the learning rate. See the following code example:

```python
#!/usr/bin/env python3
import matplotlib.pyplot as plt
import torch

# Setup optimizer + scheduler
epochs = 5
steps_per_epoch = 10
params = torch.randn(10).requires_grad_(True)
optimizer = torch.optim.SGD([params], lr=0.01, momentum=0.9)
scheduler = torch.optim.lr_scheduler.OneCycleLR(
    optimizer, max_lr=0.01, steps_per_epoch=steps_per_epoch, epochs=epochs
)

# Collect lrs
lrs = []
for step in range(epochs * steps_per_epoch):
    optimizer.step()
    scheduler.step()
    lrs.append(optimizer.param_groups[0]["lr"])

# Create plot
plt.figure()
plt.title("OneCycleLR Learning Rate Scaling")
plt.plot(lrs)
plt.yscale("log")
plt.xlabel("Steps")
plt.ylabel("Learning Rate")
plt.savefig("/tmp/OneCycleLR.png")

# Print LRs
print("Min lr: ", scheduler.optimizer.param_groups[0]["min_lr"])
print("Actual last lr: ", lrs[-1])
print("Last three lrs: ", lrs[-3:])
```

**Output**:
```
$ python test_onecyclelr.py
Min lr:  4e-08
Actual last lr:  2.0168449509684565e-05
Last three lrs:  [2.0168449509684565e-05, 4e-08, 2.0168449509684565e-05]
```

![OneCycleLR](https://user-images.githubusercontent.com/17625341/210816090-62519133-37a7-4858-a09e-ccd942481f76.png)

### Versions

```
Collecting environment information...
PyTorch version: 1.13.0+cu117
Is debug build: False
CUDA used to build PyTorch: 11.7
ROCM used to build PyTorch: N/A

OS: Fedora Linux 37 (Workstation Edition) (x86_64)
GCC version: (GCC) 12.2.1 20221121 (Red Hat 12.2.1-4)
Clang version: Could not collect
CMake version: version 3.25.1
Libc version: glibc-2.36

Python version: 3.10.8 (main, Nov 24 2022, 14:13:03) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-6.0.15-300.fc37.x86_64-x86_64-with-glibc2.36
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: Could not collect
Nvidia driver version: Could not collect
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

Versions of relevant libraries:
[pip3] numpy==1.23.5
[pip3] pytorch-lightning==1.8.3.post1
[pip3] torch==1.13.0
[pip3] torchmetrics==0.11.0
[pip3] torchvision==0.14.0
[conda] numpy                     1.23.5                   pypi_0    pypi
[conda] pytorch-lightning         1.8.3.post1              pypi_0    pypi
[conda] torch                     1.13.0                   pypi_0    pypi
[conda] torchmetrics              0.11.0                   pypi_0    pypi
[conda] torchvision               0.14.0                   pypi_0    pypi
```

cc @vincentqb @jbschlosser @albanD @janeyx99

**Comment by vadimkantorov** [comment]
@albanD It might be good to include in docs such autogenerated plots for all schedulers (similar to activation funcitons). Then such bugs would have been discovered earlier

**Comment by milesial** [comment]
I don't think this is a bug. Since the LR scheduler step is supposed to be called after the optimizer step, and it updates the groups LR, then this last LR scheduler step is updating the LR for a potential next optimizer step (which doesn't happen). The last optimizer step in your loop is taken with the low expected LR.

```
...
Stepping SGD with step 0.0012346777314112658
Adjusting learning rate of group 0 to 9.5495e-04.
Stepping SGD with step 0.0009549512084651508
Adjusting learning rate of group 0 to 7.0779e-04.
Stepping SGD with step 0.0007077932009665415
Adjusting learning rate of group 0 to 4.9519e-04.
Stepping SGD with step 0.0004951936798652628
Adjusting learning rate of group 0 to 3.1886e-04.
Stepping SGD with step 0.0003188643714987268
Adjusting learning rate of group 0 to 1.8022e-04.
Stepping SGD with step 0.00018022497577794772
Adjusting learning rate of group 0 to 8.0392e-05.
Stepping SGD with step 8.039173559862365e-05
Adjusting learning rate of group 0 to 2.0168e-05.
Stepping SGD with step 2.0168449509684565e-05
Adjusting learning rate of group 0 to 4.0000e-08.
Stepping SGD with step 4e-08
Adjusting learning rate of group 0 to 2.0168e-05.
```

If you change the loop to 

```
lrs = []
for step in range(epochs * steps_per_epoch):
    lrs.append(optimizer.param_groups[0]["lr"])
    optimizer.step()
    scheduler.step()
```
You can see the curve of the actual LRs used by the optimizers.

![image](https://user-images.githubusercontent.com/30204471/211355737-8cc8955f-3bdd-4293-a49a-cf7943d648ca.png)

**Comment by braun-steven** [comment]
Oh yes, you're absolutely right — my mistake, sorry! I will close the issue as there is nothing to be resolved and the discussion of plots in LRScheduler docs continues in #92007.
